shell/common/canvas_spy.h

```diff
@@ -0,0 +1,172 @@
+
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+#include "flutter/fml/macros.h"
+#include "third_party/skia/include/core/SkCanvas.h"
+#include "third_party/skia/include/core/SkCanvasVirtualEnforcer.h"
+#include "third_party/skia/include/utils/SkNWayCanvas.h"
+#include "third_party/skia/include/utils/SkNoDrawCanvas.h"
+
+#ifndef FLUTTER_SHELL_COMMON_CANVAS_SPY_H_
+#define FLUTTER_SHELL_COMMON_CANVAS_SPY_H_
+
+namespace flutter {
+
+class DidDrawCanvas;
+
+// Facilitates spying on drawing commands to an SkCanvas.
+//
+// This is used to determine whether anything was drawn into
+// a canvas so it is possible to implement optimizations that
+// are specific to empty canvases.
+class CanvasSpy {
+ public:
+  CanvasSpy(SkCanvas* target_canvas);
+
+  // Returns true if any non trasnparent content has been drawn into
+  // the spying canvas. Note that this class does tries to detect
+  // empty canvases but in some cases may return true even for empty
+  // canvases (e.g when a transparent image is drawn into the canvas).
+  bool DidDrawIntoCanvas();
+
+  // The returned canvas delegate all operations to the target canvas
+  // while spying on them.
+  SkCanvas* GetSpyingCanvas();
+
+ private:
+  std::unique_ptr<SkNWayCanvas> n_way_canvas_;
+  std::unique_ptr<DidDrawCanvas> did_draw_canvas_;
+
+  FML_DISALLOW_COPY_AND_ASSIGN(CanvasSpy);
+};
+
+class DidDrawCanvas : public SkCanvasVirtualEnforcer<SkNoDrawCanvas> {
+ public:
+  DidDrawCanvas(int width, int height);
+  ~DidDrawCanvas() override;
+  bool DidDrawIntoCanvas();
+
+ protected:
```

Comment:
We are not going to subclass this further. Please make this `private` and mark the class `final`. This will also make it a compiler error if the base class adds new pure virtual methods that we haven't overriden.